aten/src/ATen/core/dispatch/Dispatcher.h

```diff
@@ -287,13 +289,53 @@ class CAFFE2_API OperatorHandle final {
 
 namespace detail {
 template<class... Args> inline void unused_arg_(const Args&...) {}
+
+template <typename T, std::enable_if_t<c10::impl::not_ok_to_box<T>::value>* = nullptr>
+inline void push_ivalue_copy(std::vector<c10::IValue>& stack, const T& v) {}
+template <typename T, std::enable_if_t<!c10::impl::not_ok_to_box<T>::value>* = nullptr>
+inline void push_ivalue_copy(std::vector<c10::IValue>& stack, const T& v) {
+  torch::jit::push(stack, v);
+}
+
+template<typename Item>
+void convert_to_ivalue_vector(std::vector<c10::IValue>& stack, const Item& item) {
+  push_ivalue_copy(stack, item);
+}
+template<typename Item, typename... Rest>
+void convert_to_ivalue_vector(std::vector<c10::IValue>& stack, const Item& item, Rest... other_items) {
+  push_ivalue_copy(stack, item);
+  convert_to_ivalue_vector(stack, other_items...);
+}
+void CAFFE2_API convert_to_ivalue_vector(std::vector<c10::IValue>& stack);
 }
 
 template<class Return, class... Args>
 inline Return Dispatcher::callWithDispatchKey(const OperatorHandle& op, DispatchKey dispatchKey, Args... args) const {
   detail::unused_arg_(args...);  // workaround for a false-positive warning about unused parameters in gcc 5
   const auto& dispatchTable = op.operatorIterator_->op.dispatch_table();
   const KernelFunction& kernel = dispatch_(dispatchTable, dispatchKey);
+  // Check if we need to run callbacks registered with RecordFunction
+  // If true and callbacks need inputs, we box the arguments and pass
+  // them into the callbacks and also into the kernel call
+  at::RecordFunction guard(at::RecordScope::FUNCTION);
+  if (guard.active_) {
+    // Stack-based record function with scope lifetime can be markee as 'current'
+    // thread local record function; used to track nested recorded scopes
+    guard._setCurrent();
+    if (guard.needs_inputs_) {
+      std::vector<c10::IValue> stack;
+      detail::convert_to_ivalue_vector(stack, args...);
+
+      guard._before(op.schema().name(), stack, at::FunctionSequenceNumber::peek());
+
+      // if we could convert all the arguments, also pass the stack into the kernel call
+      if (stack.size() == sizeof...(args)) {
+        return kernel.template callUnboxedWithStack<Return, Args...>(op, stack, std::forward<Args>(args)...);
```

Comment:
sorry, I don't get why you need this call with a different codepath if the guard is active? why not just fall through and user regular call below?

Reply:
that's something that we discussed earlier with @smessmer - this is to be able to reuse the stack that we created for the observer in case we call boxAndCallBoxedFunc later